File: workers/editorial-os/src/worker/logger.ts
/**
 * Logger — 1:1 port of editorial-intelligence/workers/logger.py: run id /
 * start / finish / duration / events / errors.
 */

const LEVELS: Record<string, number> = { debug: 0, info: 1, warning: 2, error: 3 };

export interface RunLog {
  runId: string;
  startedAt: string;
  finishedAt: string | null;
  durationSeconds: number | null;
  eventsProcessed: number;
  errors: string[];
  messages: string[];
}

export function runSuccess(run: RunLog): boolean {
  return run.finishedAt !== null && run.errors.length === 0;
}

function randomRunId(): string {
  const bytes = new Uint8Array(6);
  crypto.getRandomValues(bytes);
  return [...bytes].map((b) => b.toString(16).padStart(2, "0")).join("");
}

export class WorkerLogger {
  private minLevel: number;

  constructor(minLevel: keyof typeof LEVELS = "info") {
    this.minLevel = LEVELS[minLevel] ?? 1;
  }

  start(): RunLog {
    return {
      runId: randomRunId(),
      startedAt: new Date().toISOString(),
      finishedAt: null,
      durationSeconds: null,
      eventsProcessed: 0,
      errors: [],
      messages: [],
    };
  }

  log(run: RunLog, message: string, level: keyof typeof LEVELS = "info"): void {
    if ((LEVELS[level] ?? 1) < this.minLevel) return;
    run.messages.push(`[${level}] ${message}`);
  }

  error(run: RunLog, message: string): void {
    run.errors.push(message);
    run.messages.push(`[error] ${message}`);
  }

  finish(run: RunLog, eventsProcessed: number): RunLog {
    run.finishedAt = new Date().toISOString();
    run.eventsProcessed = eventsProcessed;
    const started = Date.parse(run.startedAt);
    const finished = Date.parse(run.finishedAt);
    run.durationSeconds = (finished - started) / 1000;
    return run;
  }
}

File: editorial-intelligence/workers/logger.py
"""Worker Logger (Phase 6) — run id / start / finish / duration /
events / errors, per the spec's LOGGING section.

In-memory only; the CLI (scripts/editorial.py) decides whether/where to
persist a run's log to JSON, the same split every other Phase 1-5
persistence concern already uses (see workspace.py's own docstring).
"""
import datetime
import uuid
from dataclasses import dataclass, field
from typing import List, Optional

_LEVELS = {"debug": 0, "info": 1, "warning": 2, "error": 3}


def _now() -> str:
    return datetime.datetime.now(datetime.timezone.utc).isoformat()


@dataclass
class RunLog:
    run_id: str
    started_at: str
    finished_at: Optional[str] = None
    duration_seconds: Optional[float] = None
    events_processed: int = 0
    errors: List[str] = field(default_factory=list)
    messages: List[str] = field(default_factory=list)

    @property
    def success(self) -> bool:
        return self.finished_at is not None and not self.errors


class WorkerLogger:
    def __init__(self, min_level: str = "info"):
        self._min_level = _LEVELS.get(min_level, 1)

    def start(self) -> RunLog:
        return RunLog(run_id=uuid.uuid4().hex[:12], started_at=_now())

    def log(self, run: RunLog, message: str, level: str = "info") -> None:
        if _LEVELS.get(level, 1) < self._min_level:
            return
        run.messages.append(f"[{level}] {message}")

    def error(self, run: RunLog, message: str) -> None:
        run.errors.append(message)
        run.messages.append(f"[error] {message}")

    def finish(self, run: RunLog, events_processed: int) -> RunLog:
        run.finished_at = _now()
        run.events_processed = events_processed
        started = datetime.datetime.fromisoformat(run.started_at)
        finished = datetime.datetime.fromisoformat(run.finished_at)
        run.duration_seconds = (finished - started).total_seconds()
        return run
